### PR TITLE
Feature/FOUR-14120b: Backend - Fire job when task is assigned to a user

### PR DESCRIPTION
**Pull request by CarliPinell**
## Solution
- A Listener for the ActivityAssigned event in ProcessMaker/Providers/ProcessMakerServiceProvider.php was added in registerEvents method
- A Dispatch method for SmartInbox job was added.

## How to Test
- Create Inbox Rules
- Start a new request

## Related Tickets & Packages
https://processmaker.atlassian.net/browse/FOUR-14120

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.
ci:deploy

### CODE REVIEW

File: ProcessMaker/Jobs/SmartInbox.php [hunk 1]
```diff
@@ -8,34 +8,33 @@
...
-        $inboxRule = new InboxRule();
-        $matchingTasks = new MatchingTasks($inboxRule);
-        $matchResult = $matchingTasks->check($this->task);
-        if($matchResult){
-            //Here calls ApplyAction Class
+    {   
+        try {
+            $incomingTask = ProcessRequestToken::findOrFail($this->incomingTaskId);
+            ApplyAction::applyActionOnTask($incomingTask);
+        } catch (ModelNotFoundException $e) {
```
Comment (nolanpro) [inline ProcessMaker/Jobs/SmartInbox.php:36]: You don't need try/catch here. Let the global error handler take care of it

File: ProcessMaker/Jobs/SmartInboxExistingTasks.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        try {
+            //Load InboxRule by ID
+            $inboxRule = InboxRule::findOrFail($this->inboxRuleId);
+
+            $matchingTasks = MatchingTasks::get($inboxRule);
+            foreach ($matchingTasks as $task) {
+                ApplyAction::applyActionOnTask($task);
+            }
+        } catch (ModelNotFoundException $e) {
```
Comment (nolanpro) [inline ProcessMaker/Jobs/SmartInboxExistingTasks.php:42]: Same here

File: ProcessMaker/InboxRules/ApplyAction.php [hunk 3]
```diff
@@ -2,6 +2,66 @@
...
+        WorkflowManager::completeTask($process, $instance, $task, $data);
+    }
+
+    public function reassignToUserID($task, $inputRule)
+    {
+        $task->authorizeReassignment(Auth::user());
+        // Reassign user
+        $task->reassignTo($inputRule->reassign_to_user_id);
+        $task->persistUserData($inputRule->reassign_to_user_id);
+        $task->save();
```
Comment (nolanpro) [inline ProcessMaker/InboxRules/ApplyAction.php:60]: Looks like you don't need this, the workflow mananger in `reassignTo` takes care of saving it

File: tests/unit/ProcessMaker/InboxRules/ApplyActionTest.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        MatchingTasks::shouldReceive('matchingInboxRules')
+            ->once()
+            ->with($activeTask)
+            ->andReturn([$inboxRule]);
+
+        ApplyAction::applyActionOnTask($activeTask);
+        
+        $activeTask->refresh();
+        
+        $this->assertEquals($userToReassign->id,$activeTask->user_id);
```
Comment (nolanpro) [inline tests/unit/ProcessMaker/InboxRules/ApplyActionTest.php:130]: So it looks like `reassignTo` in ProcessMaker/Models/ProcessRequestToken.php kicks off a process to re-assign the task. So for the test, you can `WorkflowManager::shouldReceive('runProcess')->once()->with(.....) See the method in ProcessMaker/Models/ProcessRequestToken.php for the correct values. `->andReturn(12345)` any number

It also sets `$this->user_id = $res['assign_to'];` but does not save it. So remove `refresh()` and assert it equals `12345`

File: tests/unit/ProcessMaker/InboxRules/MatchingTasksTest.php [hunk 5]
```diff
@@ -75,24 +76,38 @@ public function testInboxRuleEndDate()
         // not be returned
 
         $user = User::factory()->create();
-        $completedTask = ProcessRequestToken::factory()->create([
+    
+        $task = ProcessRequestToken::factory()->create([
             'user_id' => $user->id,
-            'status' => 'COMPLETED',
+            'status' => 'ACTIVE',
```
Comment (nolanpro) [inline tests/unit/ProcessMaker/InboxRules/MatchingTasksTest.php:82]: Why change this to `ACTIVE`? The task we use for InboxRule's process_request_token_id should always be a completed task

File: tests/unit/ProcessMaker/InboxRules/MatchingTasksTest.php [hunk 6]
```diff
@@ -75,24 +76,38 @@ public function testInboxRuleEndDate()
...
-        $this->assertCount(1, $matchingTasks);
-        $this->assertEquals($inboxRule->id, $matchingTasks[0]->id);
+        $this->assertEmpty($matchingRules);
+        $this->assertNotEmpty($notMatchingRules);
+        //when end date is not past and we have a task match
+        $this->assertEquals(
+            $expectinginboxRule->process_request_token_id,
+            $notMatchingRules[0]->process_request_token_id
+        );
+
```
Comment (nolanpro) [inline tests/unit/ProcessMaker/InboxRules/MatchingTasksTest.php:111]: Try this instead. Only create 1 inbox rule for the test. First, set it to the pastEndDate and assert there are no matches. Then, $inboxRule->update to set the future end date. Then run MatchingTasks::matchingInboxRules again and assert it has 1 matching inbox rule

**Review by nolanpro — CHANGES_REQUESTED** [review]

File: ProcessMaker/InboxRules/ApplyAction.php [hunk 7]
```diff
@@ -2,6 +2,66 @@
...
+        $data = $inputRule->submit_data;
+        // Call the manager to trigger the start event
+        $process = $task->process;
+        $instance = $task->processRequest;
+        WorkflowManager::completeTask($process, $instance, $task, $data);
+    }
+
+    public function reassignToUserID($task, $inputRule)
+    {
+        $task->authorizeReassignment(Auth::user());
```
Comment (nolanpro) [inline ProcessMaker/InboxRules/ApplyAction.php:56]: Use $inboxRule->user_id here

File: ProcessMaker/Jobs/SmartInboxExistingTasks.php (hunk 2 above)
Comment (CarliPinell) [reply ProcessMaker/Jobs/SmartInboxExistingTasks.php:42]: Solved.

File: ProcessMaker/Jobs/SmartInbox.php (hunk 1 above)
Comment (CarliPinell) [reply ProcessMaker/Jobs/SmartInbox.php:36]: Solved.

File: tests/unit/ProcessMaker/InboxRules/MatchingTasksTest.php (hunk 5 above)
Comment (CarliPinell) [reply tests/unit/ProcessMaker/InboxRules/MatchingTasksTest.php:82]: My mistake, I changed the value to make some tests and I forgot to change it to COMPLETED before the PR.

File: ProcessMaker/InboxRules/ApplyAction.php (hunk 7 above)
Comment (CarliPinell) [reply ProcessMaker/InboxRules/ApplyAction.php:56]: solved

File: ProcessMaker/InboxRules/ApplyAction.php (hunk 3 above)
Comment (CarliPinell) [reply ProcessMaker/InboxRules/ApplyAction.php:60]: Removing mentioned code and leaving only `$task->reassignTo($inputRule->reassign_to_user_id);` nothing changes and it still doesn`t reassign the user_id

File: tests/unit/ProcessMaker/InboxRules/MatchingTasksTest.php (hunk 6 above)
Comment (CarliPinell) [reply tests/unit/ProcessMaker/InboxRules/MatchingTasksTest.php:111]: Solved

File: ProcessMaker/InboxRules/MatchingTasks.php [hunk 8]
```diff
@@ -8,40 +8,50 @@
...
     {
+        if (!$task || !$task->user_id) {
+            return [];
+        }
+
         $matchingInboxRules = [];
 
-        if ($task && $task->user_id) {
             //The Foreach has only inbox rules ACTIVE=true and user_id = $task->user_id
             foreach ($this->queryInboxRules($task) as $rule) {
```
Comment (nolanpro) [inline ProcessMaker/InboxRules/MatchingTasks.php:24]: Please fix the indenting here

Comment (CarliPinell) [reply ProcessMaker/InboxRules/MatchingTasks.php:24]: solved

**Pull request merged**